feat: Schema for the oldstyle edm4hep Future Circular Collider simulation Samples

prayagyadav opened the pull request:
Hi,
This is a schema for the Future Circular Collider Spring2021 pre-generated samples based on oldstyle EDM4HEP. This is a work in progress and I wanted to post it here for community inputs.
I have added most of the features: 

- [x] Added mixing classes most of the EDM4HEP classes
- [x] Association between ReconstructedParticle and MCParticle
- [x] Methods to get the daughters and parents to each MCParticle 
- [ ] Special methods for Tracks and Clusters (Need input from the community)
- [x] Tests and test root file for the FCCSchema

Further development will need inputs from more people.
Thanks

Links and References:
- Used `p8_ee_ZH_ecm240` sample from [here](https://fcc-physics-events.web.cern.ch/FCCee/delphes/spring2021/idea/) for testing (generated a smaller file with 100 events)
- [EDM4HEP](https://edm4hep.web.cern.ch/)
- [Structure of EDM4HEP files](https://github.com/HEP-FCC/FCCAnalyses/tree/master/examples/basics#structure-of-edm4hep-files)

Tagging my guides: @davidlange6 @gomber

prayagyadav committed:
Schema for the oldstyle edm4hep Future Circular Collider simulation Samples

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

prayagyadav committed:
Merge branch 'master' into fcc-schema

lgray committed:
Merge branch 'master' into fcc-schema

prayagyadav committed:
Merge branch 'CoffeaTeam:master' into fcc-schema

lgray commented on src/coffea/nanoevents/methods/fcc.py:
```diff
@@ -0,0 +1,452 @@
+            end
+        )  # force touching of the necessary data
+        return awkward.Array(
+            awkward.Array([[[]]]).layout.to_typetracer(forget_length=True)
```
This return value for the TypeTracer does not preserve the contained data type. This could lead to confusion.

It would be better to do something like:

```python3
if ak.backend(begin_end) == "typetracer":
     return index_range_numba_wrap(
         awkward.typetracer.length_one_if_typetracer(begin_end),
         awkward.ArrayBuilder(),
     ).snapshot().layout.to_typetracer(forget_length=True)
```

prayagyadav replied:
Okayy. Will check it

prayagyadav replied:
Hi, sorry for the late update. This snippet does not work. I don't really understand the logic. Could you please explain what `length_one_if_typetracer` does? Alternatively, is there a documentation/example use case for this function that I can refer to? Thanks

lgray commented on src/coffea/nanoevents/methods/fcc.py:
```diff
@@ -0,0 +1,452 @@
+        - Needs the presence of Particleidx1 collection
+        - The Particleidx1.index contains info about the daughters
+        """
+        return map_index_to_array(dask_array, dask_array.get_daughters_index, axis=2)
```
For this function and others below it you'll likely have the same problem we did with the NanoAOD schema where this will not work after cuts are placed on the object, since the indices are local to each event.

The solution we came up with was to reference back to the original array and then take only the objects that passed the original cuts. You can see an example for that flavor of GenParticles here: https://github.com/CoffeaTeam/coffea/blob/master/src/coffea/nanoevents/methods/nanoaod.py#L119

The main thing you have to do is generate the global indices upon reading in the file so that you can use those arrays instead of the local indices.

prayagyadav replied:
Hi @lgray. Sorry for the super late reply. Indeed, this won't work if cuts are already applied. But I have trouble understanding the solution you suggested. Furthermore, I don't know what `_apply_global_index` actually does. Please explain in simpler terms how I can fix this . Thanks

lgray commented on src/coffea/nanoevents/methods/fcc.py:
```diff
@@ -0,0 +1,452 @@
+    @matched_gen.dask
+    def matched_gen(self, dask_array):
+        sel = awkward.broadcast_arrays(True, dask_array)[0]
+        index = dask_array._events().MCRecoAssociations.reco_mc_index[:, :, 1]
```
Same issue with the indexing through cuts. I had most of an example together in #822.

prayagyadav committed:
Automated Commit from Fedora

prayagyadav committed:
Automated Commit from Fedora

prayagyadav committed:
parents daughters cross reference works

prayagyadav committed:
Merge branch 'CoffeaTeam:master' into review-fcc-schema

prayagyadav committed:
Merge branch 'CoffeaTeam:master' into fcc-schema

prayagyadav committed:
Cleaned

prayagyadav committed:
Merge branch 'review-fcc-schema' of https://github.com/prayagyadav/coffea into review-fcc-schema

prayagyadav committed:
Fixed all the cross references
- Added Global Indexers for parents, daughters, Muons, Electrons, MCRecoassociations
- Moved index_range function to transforms to avoid checking for typetracers etc.

lgray committed:
fix quoting issue

lgray committed:
Merge branch 'master' into fcc-schema

lgray commented on src/coffea/nanoevents/transforms.py:
```diff
@@ -141,6 +141,62 @@ def local2global(stack):
+    counts1 = awkward.num(target, axis=1)
+    counts2 = awkward.num(index, axis=1) # This is, in reality, flattened counts for axis = 2
+    
+    print(target)
```
please remove these print statements

lgray replied:
oops, that was an old commit.

lgray committed:
fix accidental deletions of .replace

prayagyadav committed:
Merge branch 'CoffeaTeam:master' into fcc-schema

prayagyadav committed:
added parents daughters tests and winter2023 sample

prayagyadav committed:
Merge branch 'fcc-schema' of https://github.com/prayagyadav/coffea into fcc-schema

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

prayagyadav committed:
Link to original samples

prayagyadav committed:
Merge branch 'fcc-schema' of https://github.com/prayagyadav/coffea into fcc-schema

lgray committed:
Merge branch 'master' into fcc-schema

prayagyadav committed:
fix mistakes in comments

lgray committed:
Merge branch 'master' into fcc-schema
